### PR TITLE
Backend: Usage of TabWidget

### PR DESCRIPTION
**Pull request by Thunderblade73**
## Dependencies
- #1150 

## What
Example usages of the Tab Widget abstraction

## Changelog Technical Details
+ Switch some stuff from TabListUpdate to TabWidgetUpdate. - Thunderblade73

### CODE REVIEW

**Comment by hannibal002** [comment]
Kuss
image

**Comment by Thunderblade73** [comment]
VisitorListenerTest is broken and idk how to fix it.

**Review by CalMWolfs — APPROVED** [review]

File: src/main/java/at/hannibal2/skyhanni/features/combat/FerocityDisplay.kt [hunk 1]
```diff
@@ -20,16 +21,18 @@ object FerocityDisplay {
...
         "combat.ferocity.tab",
-        " Ferocity: §r§c⫽(?<stat>.*)"
+        " Ferocity: §r§c⫽(?<stat>.*)",
     )
 
     private var display = ""
 
     @SubscribeEvent
-    fun onTabListUpdate(event: TabListUpdateEvent) {
+    fun onTabListUpdate(event: WidgetUpdateEvent) {
```
Comment (hannibal002) [inline src/main/java/at/hannibal2/skyhanni/features/combat/FerocityDisplay.kt:30]: why do we only check for stats widget for resetting, not for the regex?

File: src/main/java/at/hannibal2/skyhanni/features/garden/composter/ComposterDisplay.kt [hunk 2]
```diff
@@ -53,10 +54,16 @@ object ComposterDisplay {
     }
 
     @SubscribeEvent
-    fun onTabListUpdate(event: TabListUpdateEvent) {
+    fun onTabListUpdate(event: WidgetUpdateEvent) {
         if (!(config.displayEnabled && GardenAPI.inGarden())) return
+        if (!event.isWidget(TabWidget.COMPOSTER)) return
 
-        readData(event.tabList)
+        if (event.isClear()) {
```
Comment (hannibal002) [inline src/main/java/at/hannibal2/skyhanni/features/garden/composter/ComposterDisplay.kt]: removing the is clear check here will not change the logic

File: src/main/java/at/hannibal2/skyhanni/features/garden/visitor/VisitorListener.kt [hunk 3]
```diff
@@ -68,13 +69,14 @@ object VisitorListener {
     }
 
     @SubscribeEvent
-    fun onTabListUpdate(event: TabListUpdateEvent) {
+    fun onTabListUpdate(event: WidgetUpdateEvent) {
         if (!GardenAPI.inGarden()) return
+        if (!event.isWidget(TabWidget.VISITORS) && !event.isClear()) return
```
Comment (hannibal002) [inline src/main/java/at/hannibal2/skyhanni/features/garden/visitor/VisitorListener.kt]: again, the isClear check here does nothing

File: src/main/java/at/hannibal2/skyhanni/features/nether/reputationhelper/dailyquest/QuestLoader.kt [hunk 4]
```diff
@@ -36,12 +35,12 @@ class QuestLoader(private val dailyQuestHelper: DailyQuestHelper) {
         }
     }
 
-    fun loadFromTabList() {
+    fun loadFromTabList(lines: List<String>) {
```
Comment (hannibal002) [inline src/main/java/at/hannibal2/skyhanni/features/nether/reputationhelper/dailyquest/QuestLoader.kt]: Why do we request the lines as parameter? they are always the same static data, TabWidget.FACTION_QUESTS.lines

**Review by hannibal002 — CHANGES_REQUESTED** [review]
Works fine in game. Just some checks I don't think are needed.

File: src/main/java/at/hannibal2/skyhanni/features/combat/FerocityDisplay.kt (hunk 1 above)
Comment (Thunderblade73) [reply src/main/java/at/hannibal2/skyhanni/features/combat/FerocityDisplay.kt:30]: idk, what you mean.

Comment (hannibal002) [reply src/main/java/at/hannibal2/skyhanni/features/combat/FerocityDisplay.kt:30]: `event.lines.matchFirst(ferocityPattern) {`
this happens for every TabWidget update, not only for `TabWidget.STATS` and `TabWidget.DUNGEON_SKILLS_AND_STATS`

File: src/main/java/at/hannibal2/skyhanni/features/nether/reputationhelper/dailyquest/QuestLoader.kt [hunk 5]
```diff
@@ -37,11 +37,12 @@ class QuestLoader(private val dailyQuestHelper: DailyQuestHelper) {
     }
 
     fun loadFromTabList() {
+        val lines = TabWidget.FACTION_QUESTS.lines
```
Comment (hannibal002) [inline src/main/java/at/hannibal2/skyhanni/features/nether/reputationhelper/dailyquest/QuestLoader.kt]: can be inlined again

File: src/main/java/at/hannibal2/skyhanni/features/combat/FerocityDisplay.kt (hunk 1 above)
Comment (Thunderblade73) [reply src/main/java/at/hannibal2/skyhanni/features/combat/FerocityDisplay.kt:30]: oh, I missed the return

**Review by hannibal002 — CHANGES_REQUESTED** [review]

File: src/main/java/at/hannibal2/skyhanni/features/combat/FerocityDisplay.kt [hunk 6]
```diff
@@ -20,16 +21,18 @@ object FerocityDisplay {
...
         if (!isEnabled()) return
+        if (event.isWidget(TabWidget.STATS, TabWidget.DUNGEON_SKILLS_AND_STATS)) return
```
Comment (hannibal002) [inline src/main/java/at/hannibal2/skyhanni/features/combat/FerocityDisplay.kt]: reverse

**Review by hannibal002 — CHANGES_REQUESTED** [review]

**Pull request merged**